### ISSUE TITLE
Normalize text before comparing

### ISSUE THREAD

**Issue by stweil** [qurator-spk/dinglehopper#28]
UTF-8 allows different representations for the same character. Dinglehoppers currently does not detect that such different representations are identical characters, but handles them like a recognition error.

This can be fixed by normalizing the text before doing the comparison.

Example: We just had a case where the GT transcription used _zuͦſein_ (u + COMBINING RING ABOVE) while the OCR detected _zůſein_ (LATIN CAPITAL LETTER U WITH RING ABOVE). See https://digi.bib.uni-mannheim.de/fileadmin/digi/477429599/dh_055.html.

**Comment by stweil** [comment]
Dinglehopper already uses normalization (NFC). The delta in the example is caused by different characters which look the same: u + COMBINING LATIN SMALL LETTER O != u + COMBINING RING ABOVE.

So COMBINING LATIN SMALL LETTER O and COMBINING RING ABOVE should be handled as similar when comparing.

**Comment by mikegerber** [comment]
I do think that COMBINING LATIN SMALL LETTER O and COMBINING RING ABOVE are entirely different characters and not equivalent.

However,

I do think that the normalization behaviour should be configurable by the user. So if you choose to consider these characters to be the same for your use case, then you should be able to configure this. I've done some WIP on this that I need to merge and work further on it.

**Comment by cneud** [comment]
Duplicate of #11?

**Comment by stweil** [comment]
Yes, the issue can be solved with substitutions which can be configured by the users.

@mikegerber, sure, `uͦ` and `ů` are different characters, maybe like some cyrillic characters which look like latin ones. But which of them is the right one for historic German texts? Do you agree that it does not make sense to train both for OCR models in that context?

**Comment by mikegerber** [comment]
> @cneud, yes, the issue can be solved with substitutions which can be configured by the users.

Exactly.

I aim to support (Unicode terms) canonically equivalence (= using NFC consitencly) and maybe the same idea for MUFI characters. And make all further equivalence considerations user configurable. The latter also means that some of the hardcoded substitutions will be moved to some kind of configuration (#11).

**Comment by bertsky** [comment]
> > @cneud, yes, the issue can be solved with substitutions which can be configured by the users.
> 
> Exactly.

I would like to point out here that allowing _arbitrary_ equivalences also makes comparing much more difficult. Ideally, there should be sensible sets of transformations (like OCR-D GT levels) which many researchers/practitioners can agree on. And then, to facilitate commensurability, ideally, the evaluation should produce multiple metrics next to each other in the report (like: always the chosen metric plus maximum normalization (GT level 1) metric plus minimum (GT level 3 / Levenshtein) metric).

Also, cf. [existing metrics](https://github.com/ASVLeipzig/cor-asv-ann/blob/0ae6867eba39f73f5832b219f09f71788145d1c2/ocrd_cor_asv_ann/lib/alignment.py#L302-L373) in my module.

**Comment by cneud** [comment]
Basically this comes down to a number of pre-defined common use cases or scenarios, with the added possibility for users to create their own scenarios. This is also the approach that was followed during the IMPACT project and which I believe would be a sound tradeoff between comparable results and flexibility with regard to differing applications.

**Comment by mikegerber** [comment]
I agree mostly with what @bertsky and @cneud said. I just want to throw in some doubt on the belief that CERs are somehow comparable when produced by different tools. Do they count whitespace the same way? grapheme clusters? punctuation?

Side note: Is there really a set of transformations defined for OCR-D's GT level 1?

**Comment by cneud** [comment]
> the belief that CERs are somehow comparable when produced by different tools

I too strongly doubt they are! Looking at e.g. results and metrics from ICDAR papers, many resort to their own implementation for evaluation, which obviously creates considerable blur around any exact performance comparison.
